pkg/sentry/socket/plugin/cgo/socket_unsafe.go

```diff
@@ -0,0 +1,445 @@
+ssize_t plugin_recvfrom(int sockfd, void *buf, size_t len, int flags,
+			struct sockaddr *src_addr, socklen_t *addrlen, uint64_t *err);
+ssize_t plugin_recvmsg(int sockfd, struct msghdr *msg, int flags, uint64_t *err);
+ssize_t plugin_read(int fd, void *buf, size_t count, uint64_t *err);
+ssize_t plugin_readv(int fd, const struct iovec *iov, int iovcnt, uint64_t *err);
+
+// sending data operations
+ssize_t plugin_send(int sockfd, const void *buf, size_t len, int flags, uint64_t *err);
+ssize_t plugin_sendto(int sockfd, const void *buf, size_t len, int flags,
+		const struct sockaddr *dest_addr, socklen_t addrlen, uint64_t *err);
+ssize_t plugin_sendmsg(int sockfd, const struct msghdr *msg, int flags, uint64_t *err);
+ssize_t plugin_write(int fd, const void *buf, size_t count, uint64_t *err);
+ssize_t plugin_writev(int fd, const struct iovec *iov, int iovcnt, uint64_t *err);
+*/
+import "C"
+import (
+	"syscall"
+	"unsafe"
+
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	linuxerrno "gvisor.dev/gvisor/pkg/abi/linux/errno"
+)
+
+// EpollCreate works as a CGO wrapper for plugin_epoll_create.
+func EpollCreate() int {
+	return int(C.plugin_epoll_create())
+}
+
+// EpollCtl works as a CGO wrapper for plugin_epoll_ctl.
+func EpollCtl(epfd int32, op int, handle uint32, w uint64) {
+	event := syscall.EpollEvent{
+		Events: uint32(w),
+		Fd:     int32(handle),
+	}
+	C.plugin_epoll_ctl(
+		C.int(epfd),
+		C.int(op),
+		C.int(handle),
+		(*C.struct_epoll_event)(unsafe.Pointer(&event)))
+}
+
+// EpollWait works as a CGO wrapper for plugin_epoll_wait.
+func EpollWait(epfd int32, events []syscall.EpollEvent, n int, us int) int {
+	if len(events) == 0 {
+		return 0
+	}
+	return int(C.plugin_epoll_wait(
+		C.int(epfd),
+		(*C.struct_epoll_event)(unsafe.Pointer(&events[0])),
+		C.int(n),
+		C.int(us)))
+}
+
+// Socket works as a CGO wrapper for plugin_socket.
+func Socket(domain, skType, protocol int) int {
+	var errno uint64
+	if fd := int(C.plugin_socket(
+		C.int(domain),
+		C.int(skType),
+		C.int(protocol),
+		(*C.uint64_t)(unsafe.Pointer(&errno)))); fd < 0 {
+		return -int(errno)
+	} else {
+		nonblock := 1
+		C.plugin_ioctl(
+			C.int(fd),
+			(*C.uint64_t)(unsafe.Pointer(&errno)),
+			C.uint64_t(linux.FIONBIO),
+			unsafe.Pointer(&nonblock))
+		return fd
+	}
+}
+
+// Bind works as a CGO wrapper for plugin_bind.
+func Bind(handle uint32, sa []byte) int {
+	var errno uint64
+	if ret := int(C.plugin_bind(
+		C.int(handle),
+		(*C.struct_sockaddr)(GetPtr(sa)),
+		C.uint(len(sa)),
+		(*C.uint64_t)(unsafe.Pointer(&errno)))); ret != 0 {
+		return -int(errno)
+	} else {
+		return ret
+	}
+}
+
+// Listen works as a CGO wrapper for plugin_listen.
+func Listen(handle uint32, backlog int) int {
+	var errno uint64
+	if ret := int(C.plugin_listen(
+		C.int(handle),
+		C.int(backlog),
+		(*C.uint64_t)(unsafe.Pointer(&errno)))); ret != 0 {
+		return -int(errno)
+	} else {
+		return ret
+	}
+}
+
+// Accept works as a CGO wrapper for plugin_accept.
```

Comment:
Note that this also makes the socket as nonblocking, and probably note that on any other functions that do the same.

Reply:
Updated. Thx for pointing out.